### PR TITLE
Add support for anbox

### PR DESCRIPTION
**Pull request by samueldr**
~~Hey, look at the time, it's yet another rushed/incomplete port o'clock!~~

~~Hopefully this one actually works fine for me, compared to the TWRP port.~~

After some more work, it works!

There are things I'm unsure about, look at the inline comments.

Some of the commits at the tip of this branch are expected to be lopped-off before merging :). Anything below `=====` markers in the git history is assumed to be complete AFAIK.

* * *

### Things done

  - [x] Started a build
  - [x] Completed a build
  - [x] Ran a build

```nix
{ config, pkgs, lib, ... }:

{
  # x86_64
  # armv7a_neon
  device = "arm64";
  flavor = "anbox";
}
```

### CODE REVIEW

File: modules/base.nix [hunk 1]
```diff
@@ -216,9 +216,11 @@ in
     '' + (if (config.androidVersion >= 10) then ''
       echo "\$(call inherit-product-if-exists, robotnix/config/system.mk)" >> target/product/handheld_system.mk
       echo "\$(call inherit-product-if-exists, robotnix/config/product.mk)" >> target/product/handheld_product.mk
-    '' else ''
+    '' else if (config.androidVersion >= 8) /* FIXME Unclear if android 8 has these... */ then ''
```
Comment (samueldr) [inline modules/base.nix:219]: It is unclear whether something similar would be needed for Android 7.

It is unclear where Android 8 would fall here.

File: flavors/anbox/default.nix [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    #}
+    {
+      "build" = {
+        # Pre-seed mount points required for bind mounts
+        postPatch = ''
+          mkdir -p $out/blueprint
+          mkdir -p $out/kati
+          mkdir -p $out/soong
+          mkdir -p $out/make
+        '';
```
Comment (danielfullmer) [inline flavors/anbox/default.nix:45]: I have vague memories of having to do something similar with another android repo manifest.  Perhaps we could make this automatic--if we are using the bind-mount stuff and we notice one repo would be mounted inside of another one, we patch the outside repo to contain these mount points

**Comment by danielfullmer** [comment]
Currently facing this build failure:
```
[ 59% 22970/38511] target Prebuilt: webview (out/target/product/arm64/obj/APPS/webview_intermediates/package.apk)
FAILED: /bin/bash -c "(out/host/linux-x86/bin/acp -fp external/chromium-webview/prebuilt/arm64/webview.apk out/target/product/arm64/obj/APPS/webview_intermediates/package.apk ) &>
Archive:  out/target/product/arm64/obj/APPS/webview_intermediates/package.apk
  inflating: out/target/product/arm64/obj/APPS/webview_intermediates/uncompressedlibs/lib/arm64-v8a/libwebviewchromium.so  
  inflating: out/target/product/arm64/obj/APPS/webview_intermediates/uncompressedlibs/lib/armeabi-v7a/libwebviewchromium.so  
zip I/O error: Permission denied
zip error: Could not create output file (out/target/product/arm64/obj/APPS/webview_intermediates/package.apk)
```

**Comment by danielfullmer** [comment]
Another issue in the final step:
```
  $ vendor/anbox/scripts/create-package.sh /build/out/target/product/arm64/out/target/product/x86_64/ramdisk.img /build/out/target/product/arm64/out/target/product/x86_64/system.img /nix/store/f76av5r42r5xf4dkwvg2qss9rn029mxa-robotnix-anbox_arm64-2021.06.07.04/android.img
/nix/store/4y965wrv3shgg2s1ca431qpbrsb4dbag-stdenv-linux/setup: vendor/anbox/scripts/create-package.sh: /bin/bash: bad interpreter: No such file or directory
```

Could use `patchShebangs` to workaround this.  But even then, https://github.com/pmanbox/anbox/blob/master/scripts/create-package.sh looks a bit suspect--with the uses of sudo / mount / gcc.

**Comment by samueldr** [comment]
In this latest update I applied the *Jack* compilation server fixes globally, and included your fixes fixup'd when and where it made sense.

File: flavors/anbox/default.nix (hunk 2 above)
Comment (danielfullmer) [reply flavors/anbox/default.nix:37]: The workaround where you manually added mount points should be fixed by: https://github.com/danielfullmer/robotnix/pull/104

File: modules/base.nix (hunk 1 above)
Comment (danielfullmer) [reply modules/base.nix:219]: This is where we hook into the existing build system to add additional packages / robotnix configs.  For now, we don't have to support robotnix module customizations with older versions of android.

**Comment by danielfullmer** [comment]
Do you think this is this ready to merge?  You've tested that the image it produces works? We should probably write something short for the docs about it.

Did you have any ideas for further plans around building anbox images? Did you want to use robotnix to build anbox images in nixpkgs / mobile-nixos?  Like you mentioned earlier, I'm not sure we can use eval-time source fetching to pull in a version of robotnix in nixpkgs, because of restrictions in hydra.

**Comment by samueldr** [comment]
I dropped the draft because yes, I believe this is ready. Feel free to review and propose any changes.

I don't have any further idea nor plan as of now. The main goal was to give the ability to any Nix user to build their own images.

Given that we are building *basically* the same image as postmarketOS provides, as-is, there's not much benefit to providing our own build of the same thing. Though the door is open for future opportunities.

Though if *Robotnix* somehow had the ability to provide trustable build artifacts I guess that's something else to consider. But as I said no plans :).

* * *

About the documentation additions, any guidelines for what you have in mind?

**Comment by samueldr** [comment]
OH... I should actually try pre-installing F-Droid with the options as an exercise! I completely forgot about Robotnix options... I was merely interested into getting a stock build!

**Comment by samueldr** [comment]
The `microg` option failed early... Didn't look into it, but I suspect it might not support Android 7.

The F-Droid app (unsurprisingly) looks like it might work... Waiting for a build to finish.

**Comment by samueldr** [comment]
There's the slight annoyance that 24 and 25 are both for Android 7; 7.0 and 7.1.

I assumed 7.1 here.

**Comment by danielfullmer** [comment]
Thanks for the contribution!

**Pull request merged**